File: scenarios/scenario_tests/src/test/java/intellij_generate/scenarios/S03BasicTest.java
package intellij_generate.scenarios;

import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import static intellij_generate.scenarios.TestUtil.loadBazelGeneratedFile;
import static intellij_generate.scenarios.TestUtil.removeWorkingDirectory;
import static intellij_generate.scenarios.TestUtil.xpathList;
import static java.util.Arrays.asList;
import static java.util.Collections.emptyList;
import static java.util.stream.Collectors.toList;
import static org.junit.jupiter.api.Assertions.assertEquals;

public class S03BasicTest {
  private static String dolphinImlContent;
  private static String humanImlContent;
  private static String gorillaImlContent;
  private static String primateImlContent;
  private static String mammalImlContent;

  private static String modulesXmlContent;

  @BeforeAll
  public static void before_all() {
    dolphinImlContent = loadBazelGeneratedFile("03_basic/dolphin/iml.iml");
    humanImlContent = loadBazelGeneratedFile("03_basic/human/iml.iml");
    gorillaImlContent = loadBazelGeneratedFile("03_basic/gorilla/iml.iml");
    primateImlContent = loadBazelGeneratedFile("03_basic/primate/iml.iml");
    mammalImlContent = loadBazelGeneratedFile("03_basic/mammal/iml.iml");

    modulesXmlContent = loadBazelGeneratedFile("03_basic/modules_xml_modules.xml");
  }

  @Test
  public void source_folders_sample() {
    // this scenario has one source file, in a non-maven-standard layout
    assertEquals(asList(
      "file://$MODULE_DIR$/out/production/generated",
      "file://$MODULE_DIR$/out/test/generated_tests",
      "file://$MODULE_DIR$/src/main/java",
      "file://$MODULE_DIR$/src/test/java",
      "file://$MODULE_DIR$/src/main/resources",
      "file://$MODULE_DIR$/src/test/resources"),
      xpathList(dolphinImlContent, "/module/component/content/sourceFolder/@url"));

    assertEquals(
      asList("false", "true", "false", "true"),
      xpathList(dolphinImlContent, "/module/component/content/sourceFolder/@isTestSource"));
  }

  @Test
  public void sample_jar_dependencies() {
    // has no "main" dependencies
    assertEquals(emptyList(),
      xpathList(dolphinImlContent, "/module/component/orderEntry[@type='module-library' and not(@scope)]/library/CLASSES/root/@url"));

    // has standard junit 5 test jar dependencies
    assertEquals(asList(
      "external/org_junit_jupiter_junit_jupiter_api/jar/junit-jupiter-api-5.0.1.jar!/",
      "external/org_opentest4j_opentest4j/jar/opentest4j-1.0.0.jar!/"),
      removeWorkingDirectory(
        xpathList(dolphinImlContent, "/module/component/orderEntry[@type='module-library' and @scope='TEST']/library/CLASSES/root/@url")));
  }

  @Test
  public void module_dependencies() {
    assertEquals(emptyList(),
      xpathList(mammalImlContent, "/module/component/orderEntry[@type='module']/@module-name"));

    assertEquals(asList("mammal"),
      xpathList(primateImlContent, "/module/component/orderEntry[@type='module']/@module-name"));

    assertEquals(asList("mammal"),
      xpathList(dolphinImlContent, "/module/component/orderEntry[@type='module']/@module-name"));

    assertEquals(asList("mammal", "primate"),
      xpathList(humanImlContent, "/module/component/orderEntry[@type='module']/@module-name"));

    assertEquals(asList("mammal", "primate"),
      xpathList(gorillaImlContent, "/module/component/orderEntry[@type='module']/@module-name"));
  }

  @Test
  public void modules_xml() {
    assertEquals(emptyList(),
      xpathList(modulesXmlContent, "/project/component/modules/module/fileurl").stream()
        .map(u -> u.split("/")[u.split("/").length-1])
        .collect(toList()));
  }
}

File: scenarios/scenario_tests/pytest/s03_basic_test.py
import unittest
from scenario_test_support import *

class S03BasicTest(unittest.TestCase):

    archive = load_archive("03_basic/intellij_files")

    dolphin_iml_content = archive["03_basic/dolphin/dolphin.iml"]
    human_iml_content = archive["03_basic/human/human.iml"]
    gorilla_iml_content = archive["03_basic/gorilla/gorilla.iml"]
    primate_iml_content = archive["03_basic/primate/primate.iml"]
    mammal_iml_content = archive["03_basic/mammal/mammal.iml"]

    modules_xml_content = archive["03_basic/.idea/modules.xml"]

    def test_source_folders_sample(self):
        self.assertEqual([
            "file://$MODULE_DIR$/out/production/generated",
            "file://$MODULE_DIR$/out/test/generated_tests",
            "file://$MODULE_DIR$/src/main/java",
            "file://$MODULE_DIR$/src/test/java"
        ], xpath_attribute_list(self.dolphin_iml_content, "./component/content/sourceFolder", "url"))

        self.assertEqual(
            ["false", "true", "false", "true"],
            xpath_attribute_list(self.dolphin_iml_content, "./component/content/sourceFolder", "isTestSource"))

    def test_output_folders(self):
        self.assertEqual(
            ["file://$MODULE_DIR$/out/production"],
            xpath_attribute_list(self.dolphin_iml_content, "./component/output", "url"))

        self.assertEqual(
            ["file://$MODULE_DIR$/out/test"],
            xpath_attribute_list(self.dolphin_iml_content, "./component/output-test", "url"))

    def test_libraries(self):
        self.assertEqual([], find_all_plain_jar_libraries(self.dolphin_iml_content))
        self.assertEqual(junit5_jars(), find_all_test_jar_libraries(self.dolphin_iml_content))

    def test_module_dependencies(self):
        self.assertEqual([],
                         xpath_attribute_list(self.mammal_iml_content,
                                              "./component/orderEntry[@type='module']",
                                              "module-name"))

        self.assertEqual(["mammal"],
                         xpath_attribute_list(self.primate_iml_content,
                                              "./component/orderEntry[@type='module']",
                                              "module-name"))

        self.assertEqual(["mammal"],
                         xpath_attribute_list(self.dolphin_iml_content,
                                              "./component/orderEntry[@type='module']",
                                              "module-name"))

        self.assertEqual(["mammal", "primate"],
                         xpath_attribute_list(self.human_iml_content,
                                              "./component/orderEntry[@type='module']",
                                              "module-name"))

        self.assertEqual(["mammal", "primate"],
                         xpath_attribute_list(self.gorilla_iml_content,
                                              "./component/orderEntry[@type='module']",
                                              "module-name"))

    def test_modules_xml(self):
        self.assertEqual([
            "file://$PROJECT_DIR$/dolphin/dolphin.iml",
            "file://$PROJECT_DIR$/gorilla/gorilla.iml",
            "file://$PROJECT_DIR$/human/human.iml",
            "file://$PROJECT_DIR$/mammal/mammal.iml",
            "file://$PROJECT_DIR$/primate/primate.iml",
        ], xpath_attribute_list(self.modules_xml_content, "./component/modules/module", "fileurl"))
